client/verta/verta/_pipeline_orchestrator/_orchestrator.py

```diff
@@ -42,14 +61,80 @@ def _prepare_pipeline(self) -> TopologicalSorter:
             If the pipeline graph has cycles.
 
         """
-        dag = TopologicalSorter(self._pipeline_spec["graph"]["inputs"])
+        dag = TopologicalSorter(self._get_steps_input_names(self._pipeline_spec))
         dag.prepare()
         # TODO: assert one input node
         # TODO: assert one output node
 
         self._dag = dag
         self._outputs = dict()
 
+    def _get_step_input(
+        self,
+        name: str,
+    ) -> Any:
+        """Get a step's input values from ``self._outputs``.
+
+        Parameters
+        ----------
+        name : str
+            Step name.
+
+        Returns
+        -------
+        ValueError
+            If `name` has no input steps.
+        RuntimeError
+            If `name`'s input steps are not in ``self._outputs``.
+
+        """
+        predecessors: Set[str] = self._step_handlers[name].predecessors
+        if not predecessors:
+            raise ValueError(
+                f"unexpected error: step {name} has no input steps",
+            )
+        if not predecessors <= self._outputs.keys():
+            raise RuntimeError(
+                f"unexpected error: step {name}'s input steps' outputs not found",
+            )
+
+        # TODO: figure out how we actually want to collect upstream outputs
+        if len(predecessors) == 1:
+            return self._outputs[list(predecessors)[0]]
+        else:
+            # TODO: figure out how to orchestrate complex pipelines
+            raise ValueError("multiple inputs not yet supported")
+            # input = {
+            #     predecessor: self._outputs[predecessor]
+            #     for predecessor in predecessors
+            # }
```

Comment:
This is pulled directly out of `_run_step()` (with minor changes to exception messages).